specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,391 @@
+| `_mint` | `uint256` | The amount of token to deposit |
+| `_value` | `uint256` | The value of the deposit transaction, used to transfer native asset that is already on L2 from L1 |
+| `_gasLimit` | `uint64` | The gas limit of the deposit transaction |
+| `_isCreation` | `bool` | Signifies the `_data` should be used with `CREATE` |
+| `_data` | `bytes` | The calldata of the deposit transaction |
+
+#### `depositTransaction`
+
+This function is the hot code path for all usage of the `L1StandardBridge` and `L1CrossDomainMessenger`, so it MUST
+be as backwards compatible as possible. It MUST revert for chains using custom gas token when `ether` is sent with
+the `CALL`. This prevents `ether` from being stuck in the `OptimismPortal`, since it cannot be used to mint native
+asset on L2.
+
+#### `setGasPayingToken`
+
+This function MUST only be callable by the `SystemConfig`. When called, it creates a special deposit transaction
+from the `DEPOSITOR_ACCOUNT` that calls the `L1Block.setGasPayingToken` function.
+
+```solidity
+function setGasPayingToken(address _token, uint8 _decimals) external;
+```
+
+### StandardBridge
+
+The `StandardBridge` contracts on both L1 and L2 MUST be aware of the custom gas token. The `ether` specific ABI on the
+`StandardBridge` is disabled when custom gas token is enabled.
+
+The following methods MUST revert when `CALLVALUE` is non zero:
+
+- `bridgeETH(uint32,bytes)`
+- `bridgeETHTo(address,uint32,bytes)`
+- `receive()`
+- `finalizeBridgeETH(address, address, uint256, bytes)`
+
+The following legacy methods MUST also revert when `CALLVALUE` is non zero:
+
+- `depositETH(uint32,bytes)`
+- `depositETHTo(address,uint32,bytes)`
+- `finalizeETHWithdrawal(address,address,uint256,bytes)`
+- `withdraw(address,uint256,uint32,bytes)`
+- `withdrawTo(address,address,uint256,uint32,bytes)`
+- `finalizeDeposit(address,address,address,address,uint256,bytes)`
+
+### CrossDomainMessenger
+
+The `CrossDomainMessenger` contracts on both L1 and L2 MUST not be able to faciliate the transfer of native asset
+on custom gas token chains due to their tight coupling with `ether` and the `OptimismPortal` and `L2ToL1MessagePasser`
+contracts.
+
+It is possible to support this in the future with a redesign of the `CrossDomainMessenger` contract.
+
+The following methods MUST revert when `CALLVALUE` is non zero:
+
+- `sendMessage(address,bytes,uint32)`
+- `relayMessage(uint256,address,address,uint256,uint256,bytes)`
+
+It MAY be safe to not always directly revert in these implementations if the revert happens elsewhere in the callstack.
+
+### SystemConfig
+
+The `SystemConfig` is modified to allow the address of the custom gas paying token to be set during the call
+to `initialize`. When the custom gas token is set, `OptimismPortal.setGasPayingToken` is called. The implementation
+of `initialize` MUST not allow the chain operator to change the address of the custom gas token if it is already set.
```

Comment:
instead of relying on `SystemConfig` making a call to `OptimismPortal.setGasPayingToken`, could `OptimismPortal` read whether a custom gas token is being used by `SystemConfig` and just its behavior automatically? this would allow the contract to internalize the logic inside existing methods (instead of creating new methods to separate logic). no strong feelings about this -- just curious

Reply:
The `setGasPayingToken` function here emits a special deposit tx that sets it in the `L1Block` contract

Reply:
The `OptimismPortal` does call to the `SystemConfig` as the source of truth